src/openforms/js/components/admin/form_design/utils.js

```diff
@@ -57,6 +57,74 @@ const checkKeyChange = (mutationType, newComponent, oldComponent) => {
   return newComponent.key !== oldComponent.key;
 };
 
+const checkIfInitialValueTransformationIsNeeded = (newType, originalVariable) => {
+  // If the initialValue is null/undefined, we should transform
+  if (originalVariable?.initialValue == null) {
+    return true;
+  }
+
+  switch (newType) {
+    case 'array':
+      // If the value isn't in array format, it should be transformd
+      return !Array.isArray(originalVariable.initialValue);
+
+    case 'boolean':
+      return typeof originalVariable.initialValue !== 'boolean';
+
+    case 'object':
+      return typeof originalVariable.initialValue !== 'object';
+
+    case 'int':
+    case 'float':
+      return typeof originalVariable.initialValue !== 'number';
+
+    case 'string':
+    case 'datetime':
+    case 'date':
+    case 'time':
+      return typeof originalVariable.initialValue !== 'string';
+
+    default:
+      return false;
+  }
+};
+
+const transformInitialValue = (newType, originalVariable) => {
+  switch (newType) {
+    case 'array':
+      if (
+        originalVariable?.initialValue == null ||
+        typeof originalVariable.initialValue === 'object'
+      ) {
+        return [];
+      }
+      return [originalVariable.initialValue];
+
+    case 'boolean':
+      return undefined;
+
+    case 'int':
+      return Number.parseInt(originalVariable.initialValue);
```

Comment:
if `initialValue` is `'a'`, this will result in NaN and then we have an entirely different class of problems. I'm in favour of just always resetting to default empty value when you change type.

Reply:
Yeah for all dataType's the transformation now just resets the value. So array becomes `[]`, object becomes `{}`, bool, int, and float return to `undefined`, and string, date, datetime and time return to `''`.